docs/hub/security-pickle.md

```diff
@@ -0,0 +1,226 @@
+```
+
+We can see that there isn’t much in there, a few opcodes and the associated data. You might be thinking, so what’s the problem with pickle ?
+
+Let’s try something else :
+
+```python
+from fickling.pickle import Pickled
+import pickle
+
+# Create a malicious pickle
+data = "my friend needs to know this"
+
+pickle_bin = pickle.dumps(data)
+
+p = Pickled.load(pickle_bin)
+
+p.insert_python_exec('print("you\'ve been pwned !")')
+
+with open('payload.pkl', 'wb') as f:
+    p.dump(f)
+
+# innocently unpickle and get your friend's data
+with open('payload.pkl', 'rb') as f:
+    data = pickle.load(f)
+		print(data)
+```
+
+Here we’re using the [fickling](https://github.com/trailofbits/fickling) library for simplicity. It allows us to add pickle instructions to execute code contained in a string via the `exec` function. This is how you circumvent the fact that you cannot define functions or classes in your pickles : you run exec on python code saved as a string.
+
+When you run this, it creates a `payload.pkl` and prints the following :
+
+```python
+you've been pwned !
+my friend needs to know this
+```
+
+If we check the contents of the pickle file we get :
+
+```python
+# cat payload.pkl
+c__builtin__
+exec
+(Vprint("you've been pwned !")
+tR my friend needs to know this.%
+
+# hexyl payload.pkl
+┌────────┬─────────────────────────┬─────────────────────────┬────────┬────────┐
+│00000000│ 63 5f 5f 62 75 69 6c 74 ┊ 69 6e 5f 5f 0a 65 78 65 │c__built┊in___exe│
+│00000010│ 63 0a 28 56 70 72 69 6e ┊ 74 28 22 79 6f 75 27 76 │c_(Vprin┊t("you'v│
+│00000020│ 65 20 62 65 65 6e 20 70 ┊ 77 6e 65 64 20 21 22 29 │e been p┊wned !")│
+│00000030│ 0a 74 52 80 04 95 20 00 ┊ 00 00 00 00 00 00 8c 1c │_tR×•× 0┊000000×•│
+│00000040│ 6d 79 20 66 72 69 65 6e ┊ 64 20 6e 65 65 64 73 20 │my frien┊d needs │
+│00000050│ 74 6f 20 6b 6e 6f 77 20 ┊ 74 68 69 73 94 2e       │to know ┊this×.  │
+└────────┴─────────────────────────┴─────────────────────────┴────────┴────────┘
+```
+
+Basically, this is what’s happening when you unpickle :
+
+```python
+# ...
+opcodes_stack = [exec_func, "malicious argument", "REDUCE"]
+opcode = stack.pop()
+if opcode == "REDUCE":
+    arg = opcodes_stack.pop()
+    callable = opcodes_stack.pop()
+    opcodes_stack.append(callable(arg))
+# ...
+```
+
+The instructions that pose a threat are `STACK_GLOBAL`, `GLOBAL` and `REDUCE`.
+
+`REDUCE` is what tells the unpickler to execute the function with the provided arguments and `*GLOBAL` instructions are telling the unpickler to `import` stuff.
+
+To sum up, pickle is dangerous because :
+
+- when importing a python module, arbitrary code can be executed
+- you can import builtin functions like `eval` or `exec`, which can be used to execute arbitrary code
+- when instantiating an object, the constructor may be called
+
+This is why it is stated in most docs using pickle, do not unpickle data from untrusted sources.
+
+## Mitigation Strategies
```

Comment:
can we move the mitigation strategies and recommandations to the top? Feels like most people will want a TL;DR on this

Reply:
I've added a TOC and a summary at the top. Does this work better or do you want the mitigation strategies at the top ?

Reply:
works for me to add this as a TL;DR like you did, but i would still allude to the main short-term mitigations

Reply:
(see my suggestion)